Add configurations for userns

avisiedo opened the pull request:
- (removed) Add configuration for updating runc rpm package.
- (removed) Add configuration for updating cri-o rpm package.
- Add configuration for updating rpm packages.
- Add configuration for subids.
- Add a userns composition for creating all the configs
  with one command.
- Update configuration for cgroup-v2.

Changes based on [Demo: namespaced systemd workloads on OpenShift](https://frasertweedale.github.io/blog-redhat/posts/2021-07-22-openshift-systemd-workload-demo.html).

Thanks to @frasertweedale

avisiedo committed:
Add configurations for userns:

- Add configuration for updating runc rpm package.
- Add configuration for updating cri-o rpm package.
- Add configuration for subids.
- Add a userns composition for creating all the configs
  with one command.
- Update configuration for cgroup-v2.

frasertweedale commented on config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst:
```diff
@@ -0,0 +1,30 @@
+
+          [Service]
+          Type=oneshot
+          ExecStart=bash -c 'rpm -q "$( rpm -q ${RPM_CRI_O} )" || rpm-ostree override replace --reboot "${RPM_CRI_O}"'
```
Side note: we are usually overriding `runc`, not `cri-o` :)

From this command construction, I assume that we were unable to get it to work with only
`rpm-ostree override replace --idempotent --reboot`.  That is, was it unconditionally rebooting?
And is that why we must explicitly test the package version via `rpm` program?

I don't like that the package must be unconditionally downloaded upon every reboot, in order to perform
this test.  Perhaps two separate variables are better - one for the package name and the other for the
URL?  The `Makefile` can even take care of this, deriving the package name from the RPM and setting a
variable - rather than this being done by systemd on the worker node, every time it boots.

avisiedo replied:
about cri-o: Yea, I know! but the effort was a little to get the config for cri-o too, so we have the automation just in case.
about the command: I was using previously`'rpm-ostree install --idempotent --reboot`, but I should have tested wrongly; when I was doing final tests I saw that it was not installing the package and I searched for a workaround.

Thanks @frasertweedale for point out the uncondicional download, I will update the PR for using two variables, so it is downloaded only when it is needed.

frasertweedale replied:
I think the new approach is good, avoiding the unconditional download during bootup and only downloading it if the desired RPM is not installed. :+1:

frasertweedale commented on config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst:
```diff
@@ -0,0 +1,30 @@
+          [Unit]
+          Description=Install custom cri-o
+          # https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/
+          After=network-online.target
```
This unit does not work for me on OCP 4.10.  Logs:
```
% oc adm node-logs node/ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx -u rpm-overrides.service                                                                    
-- Logs begin at Thu 2021-12-02 02:52:45 UTC, end at Thu 2021-12-02 05:20:26 UTC. --
-- No entries --
-- Logs begin at Thu 2021-12-02 02:52:45 UTC, end at Thu 2021-12-02 05:20:26 UTC. --
Dec 02 05:00:24.321949 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: Starting Install RPM overrides...
Dec 02 05:00:24.339044 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: package runc-1.0.2-992.rhaos4.10.el8.x86_64 is not installed
Dec 02 05:00:25.177411 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Downloading 'https://ftweedal.fedorapeople.org/runc-1.0.2-992.rhaos4.10.el8.x86_64.rpm'... done!
Dec 02 05:00:29.025484 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Checking out tree d95b9a4...done
Dec 02 05:00:29.089007 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: No enabled rpm-md repositories.
Dec 02 05:00:29.203517 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Importing rpm-md...done
Dec 02 05:00:29.212301 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Resolving dependencies...done
Dec 02 05:00:29.214332 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Applying 1 override
Dec 02 05:00:29.306524 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Processing packages...done
Dec 02 05:00:29.309166 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Running pre scripts...done
Dec 02 05:00:29.310469 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Running post scripts...done
Dec 02 05:00:30.232707 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Running posttrans scripts...done
Dec 02 05:00:30.420456 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Writing rpmdb...done
Dec 02 05:00:32.387732 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Writing OSTree commit...done
Dec 02 05:00:35.136481 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Staging deployment...done
Dec 02 05:00:35.137815 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: error: Bus owner changed, aborting. This likely means the daemon crashed; check logs with `journalctl -xe`.
Dec 02 05:00:35.141900 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: rpm-overrides.service: Main process exited, code=exited, status=1/FAILURE
Dec 02 05:00:35.142056 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: rpm-overrides.service: Failed with result 'exit-code'.
Dec 02 05:00:35.142501 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: Failed to start Install RPM overrides.
```

Perhaps it is being executed too early in boot sequence?  Perhaps it needs to run `After=rpm-ostreed.service`?

avisiedo replied:
Thanks @frasertweedale ; I tested on OCP 4.9.5; I am installing an OCP 4.10 for fixing this.

avisiedo replied:
I have tried in OCP 4.9.9 and it was successful too; the output I get is the below:

```raw
-- Logs begin at Fri 2021-11-26 04:40:34 UTC, end at Thu 2021-12-02 21:08:58 UTC. --
Dec 02 21:07:01 crc-x4qnm-master-0 bash[1663]: package runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64 is not installed
Dec 02 21:07:01 crc-x4qnm-master-0 systemd[1]: Starting Install custom runc...
Dec 02 21:07:05 crc-x4qnm-master-0 bash[1663]: Downloading 'https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm'... done!
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Checking out tree b690184...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: No enabled rpm-md repositories.
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Importing rpm-md...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Resolving dependencies...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Applying 1 override and 5 overlays
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Processing packages...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Running pre scripts...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Running post scripts...done
Dec 02 21:07:08 crc-x4qnm-master-0 bash[1663]: Running posttrans scripts...done
Dec 02 21:07:08 crc-x4qnm-master-0 bash[1663]: Writing rpmdb...done
Dec 02 21:07:09 crc-x4qnm-master-0 bash[1663]: Writing OSTree commit...done
Dec 02 21:07:10 crc-x4qnm-master-0 bash[1663]: Staging deployment...done
Dec 02 21:07:11 crc-x4qnm-master-0 systemd[1]: install-runc.service: Succeeded.
Dec 02 21:07:11 crc-x4qnm-master-0 systemd[1]: Stopped Install custom runc.
Dec 02 21:07:11 crc-x4qnm-master-0 systemd[1]: install-runc.service: Consumed 161ms CPU time
-- Reboot --
Dec 02 21:08:00 crc-x4qnm-master-0 bash[1648]: runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: Starting Install custom runc...
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: install-runc.service: Succeeded.
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: Started Install custom runc.
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: install-runc.service: Consumed 15ms CPU time
```
Comparing with the output that you get, I can see an error message at the end; the message indicate the daemon crashed; I want to try in OCP 4.10 to dig more on this and replay it. I am having some situations to get the cluster ready.

By the way, your traces inspired me for getting a more generalized overlay for installing the rpm packages, thank you; I have updated the PR so that now we have only one overlay that can be customized for the POOL and the list of packages. I have tested in CodeReadyContainers so far.

frasertweedale replied:
Unfortunately, I get the exact same failure on OCP 4.9 (nightly) via cluster-bot.  Must be something different about how nodes are managed in CRC.

Recommend you install cluster(s) on PSI or via cluster-bot for testing.

avisiedo replied:
Thanks, I replayed the situation in OCP 4.9.9 in PSI.

tiran replied:
``network-online`` does not meant that you can make network connections and lookup hostnames in DNS. You have to wait until after name lookups work:

```
After=network-online.target nss-lookup.target
Wants=network-online.target
```

tiran replied:
https://www.freedesktop.org/software/systemd/man/systemd.special.html

> A target that should be used as synchronization point for all host/network name service lookups.

avisiedo replied:
thanks @tiran for the comments and the link to the documentation; updating the systemd unit generated.

avisiedo committed:
Update for generating the list of resources depending on POOL variables

avisiedo committed:
Update for generating the list of resources depending on POOL variables

avisiedo committed:
Updates by the code review and improvements:

- Add rpm-overrides overlay that generalize the rpm
  package overrides for the nodes.
- Update userns for using rpm-overrides overlay.
- Update .gitignore file.
- Remove cri-o and runc overlays as they are generalized
  by rpm-overrides overlay.

avisiedo committed:
Fix rpm could fail when installing the rpm

frasertweedale commented on config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst:
```diff
@@ -20,11 +20,15 @@ spec:
-          After=network-online.target rpm-ostreed.service
+          After=network-online.target
+          # After=network-online.target
+          StartLimitIntervalSec=10
+          StartLimitBurst=30
```
I don't think these are really required... or if we do decide we want it (e.g. prevent unlimited download of external file due to unlimited restart), then these values are not going to achieve the desired effect.

From `systemd.unit(5)`, these settings work together to inhibit further restarts if the service restarts *burst* times in *interval-sec* seconds.  But here we are downloading an external RPM, so it seems unlikely that we would ever hit the rate of 30 in 10 seconds (or 3 per second).

So I would suggest either remove the setting (it is for a hack after all, not for producion use), or else change these settings to "10 times in 5 minutes" or something like that.  Alternatively, under `[Service]` specify a high `RestartSec` to configure a longer delay between restart attempts (perhaps 1 minute).

avisiedo replied:
Thanks @frasertweedale for your comments. I misread the systemd documentation and I wanted to set the values for 30 max intents to restart and a delay of 10 seconds between the retries (my thoughts).

I like "10 times in 5 minutes". I will change it. I want to limit this to avoid the systemd.unit is restarting the systemd.unit if for some reason the command fails continuously. About RestartSec I will set this to 10 seconds, so that if the process fails, we don't have to await too much until the RPM package is tried to be installed again.

I have seen in PSI cluster that first time is always failing. I am trying to retrieve a traceback that I saw the last Friday, so it could point out directly to the root cause.

In short: If the systemd.unit is failing continuously, it will be restarted with a delay of 10 seconds from the last failure; moreover the systemd.unit will stop restarting if the threshold is reached for 10 restarts in 5 minutes, avoiding that this is restarting and downloading the RPM packages continuously.

frasertweedale replied:
Thanks, the new values LGTM.   I think we can do a final review during collab review session today, then merge.

frasertweedale commented on config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst:
```diff
@@ -20,11 +20,15 @@ spec:
 
           [Service]
-          Type=oneshot
+          Type=simple
```
[nit] we can remove this line; `simple` is the default when specifying `ExecStart`.

avisiedo replied:
Thanks, removing and keeping it cleaner.

frasertweedale reviewed (dismissed):
It works, good stuff.  A couple of nits (see comments).  Feel free to amend, or merge it as is.

avisiedo committed:
Add RestartSec and update StartLimitIntervalSec and StartLimitBurst systemd attributes

avisiedo committed:
Remove Type=simple from systemd definition

avisiedo committed:
Fix documentation for rpm-overrides overlay configuration

avisiedo committed:
Make available the dns resolution for the system.unit defined

avisiedo commented on config/static/nodes/cgroup-v2/kustomization.yaml:
```diff
@@ -2,5 +2,3 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
```
- Remove this file from the repository.
- Add this path to the .gitignore contents.
- Generate the file from the Makefile rule.

avisiedo commented on config/static/nodes/cgroup-v2/Makefile:
```diff
@@ -0,0 +1,6 @@
+
+.PHONY: configure
+configure:
+	@git checkout kustomization.yaml
```
- Remove checkout, and generate the head of the file.

avisiedo commented on config/static/nodes/rpm-overrides/Makefile:
```diff
@@ -0,0 +1,26 @@
+POOL ?= worker master
+RPM_PACKAGES ?=
+# RPM_RUNC ?= https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm
+# RPM_RUNC_PKG_NAME ?= runc-1.0.0-990.rhaos4.8.gitcd80260.el8
```
Cleanup commented lines

avisiedo commented on config/static/nodes/rpm-overrides/Makefile:
```diff
@@ -0,0 +1,26 @@
+POOL ?= worker master
+RPM_PACKAGES ?=
```
Add description for the variables

avisiedo commented on config/static/nodes/rpm-overrides/Makefile:
```diff
@@ -0,0 +1,26 @@
+.PHONY: configure
+configure:
+ifneq (,$(RPM_PACKAGES))
+	git checkout kustomization.yaml
```
The same as before; do not checkout this file, and generate the head from here, and delete kustomize.yaml from repository and don't checkout it.

avisiedo commented on config/static/nodes/rpm-overrides/README.md:
````diff
@@ -0,0 +1,56 @@
+Basic usage:
+
+```shell
+export RPM_PACKAGES="https://..."
````
Add description an example about what is expected on this variable.

avisiedo commented on config/static/nodes/rpm-overrides/README.md:
````diff
@@ -0,0 +1,56 @@
+
+```shell
+export RPM_PACKAGES="https://..."
+make -C config/static/nodes/runc configure
+kustomize build config/static/nodes/runc | oc create -f -
````
Wrong reference into the paths, it should be: ` config/static/nodes/rpm-overrides`.

avisiedo commented on config/static/nodes/rpm-overrides/README.md:
````diff
@@ -0,0 +1,56 @@
+```shell
+export RPM_PACKAGES="https://..."
+export POOL="master"
+make -C config/static/nodes/runc configure
+kustomize build config/static/nodes/runc | oc create -f -
````
Wrong reference into the paths, it should be: `config/static/nodes/rpm-overrides`.

avisiedo commented on config/static/nodes/rpm-overrides/README.md:
````diff
@@ -0,0 +1,56 @@
+The below install a custom runc package into all the nodes.
+
+```shell
+POOL="master worker" RPM_PACKAGES=https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm make configure
````
```suggestion
POOL="master worker" RPM_PACKAGES="https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm" make configure
```

avisiedo commented on config/static/nodes/rpm-overrides/kustomization.yaml:
```diff
@@ -0,0 +1,5 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Generate the head of this file into the Makefile, remove this file from the repository, and add this file to .gitignore file.

avisiedo commented on config/static/nodes/userns-subid/Makefile:
```diff
@@ -0,0 +1,17 @@
+
+.PHONY: configure
+configure:
+	git checkout kustomization.yaml
```
Remove this checkout operation.

avisiedo commented on config/static/nodes/userns-subid/kustomization.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
The same as the previous ones.

avisiedo committed:
Update pull request review changes:

- Remove kustomization.yaml and regenerate from Makefile rule for
  rpm-overrides overlay.
- Document the expected content for RPM_PACKAGES and show examples.
- Fix wrong paths in the rpm-overrides overlay.
- Remove kustomization.yaml from userns-subid overlay and generate
  it from Makefile file.
- Update .gitignore for the generated files for the 'nodes' overlays.

avisiedo committed:
Remove non existing paths from .gitignore

avisiedo committed:
Update documentation for rpm-overrides overlay

avisiedo committed:
Add Fraser's blog references to the README.md files

avisiedo committed:
Add one missed reference to Fraser's blog entry

avisiedo committed:
Fix commands in README.md file

avisiedo committed:
Fix comment generated to retrieve log information about the rpm installation
